// @sapphire/plugin-subcommands/dist/cjs/index.cjs
'use strict';

var PluginSubcommandCooldown_cjs = require('./preconditions/PluginSubcommandCooldown.cjs');
var Subcommand_cjs = require('./lib/Subcommand.cjs');
var subcommandCooldown_cjs = require('./lib/precondition-resolvers/subcommandCooldown.cjs');
var Enums_cjs = require('./lib/types/Enums.cjs');
var Events_cjs = require('./lib/types/Events.cjs');
var SubcommandMappings_cjs = require('./lib/types/SubcommandMappings.cjs');
var _load_cjs = require('./listeners/_load.cjs');
var _load_cjs$1 = require('./preconditions/_load.cjs');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var subcommandCooldown_cjs__namespace = /*#__PURE__*/_interopNamespace(subcommandCooldown_cjs);

// src/index.ts
var SubcommandPreconditions = {
  PluginSubcommandCooldown: PluginSubcommandCooldown_cjs.PluginPrecondition
};
var version = "6.0.3";

exports.SubcommandPreconditionResolvers = subcommandCooldown_cjs__namespace;
Object.defineProperty(exports, "loadListeners", {
  enumerable: true,
  get: function () { return _load_cjs.loadListeners; }
});
Object.defineProperty(exports, "loadPreconditions", {
  enumerable: true,
  get: function () { return _load_cjs$1.loadPreconditions; }
});
exports.SubcommandPreconditions = SubcommandPreconditions;
exports.version = version;
Object.keys(Subcommand_cjs).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return Subcommand_cjs[k]; }
  });
});
Object.keys(Enums_cjs).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return Enums_cjs[k]; }
  });
});
Object.keys(Events_cjs).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return Events_cjs[k]; }
  });
});
Object.keys(SubcommandMappings_cjs).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return SubcommandMappings_cjs[k]; }
  });
});
