// @sapphire/plugin-subcommands/dist/cjs/preconditions/PluginSubcommandCooldown.cjs
'use strict';

var framework = require('@sapphire/framework');
var ratelimits = require('@sapphire/ratelimits');
var discord_js = require('discord.js');
require('../lib/Subcommand.cjs');
var Enums_cjs = require('../lib/types/Enums.cjs');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
var _PluginPrecondition = class _PluginPrecondition extends framework.AllFlowsPrecondition {
  constructor() {
    super(...arguments);
    __publicField(this, "subcommandBuckets", /* @__PURE__ */ new WeakMap());
  }
  messageRun(message, subcommand, context) {
    const cooldownId = this.getIdFromMessage(message, context);
    return this.sharedRun(message.author.id, subcommand, context, cooldownId, "message");
  }
  chatInputRun(interaction, subcommand, context) {
    const cooldownId = this.getIdFromInteraction(interaction, context);
    return this.sharedRun(interaction.user.id, subcommand, context, cooldownId, "chat input");
  }
  contextMenuRun(interaction, subcommand, context) {
    const cooldownId = this.getIdFromInteraction(interaction, context);
    return this.sharedRun(interaction.user.id, subcommand, context, cooldownId, "context menu");
  }
  sharedRun(authorId, subcommand, context, cooldownId, commandType) {
    if (context.external)
      return this.ok();
    if (!context.delay)
      return this.ok();
    if (context.filteredUsers?.includes(authorId))
      return this.ok();
    const rateLimit = this.getManager(subcommand, context).acquire(cooldownId);
    if (rateLimit.limited) {
      const remaining = rateLimit.remainingTime;
      const nextAvailable = discord_js.time(Math.floor(rateLimit.expires / 1e3), discord_js.TimestampStyles.RelativeTime);
      return this.error({
        identifier: Enums_cjs.SubcommandIdentifiers.SubcommandPreconditionCooldown,
        message: `There is a cooldown in effect for this ${commandType} subcommand. It'll be available ${nextAvailable}.`,
        context: { remaining }
      });
    }
    rateLimit.consume();
    return this.ok();
  }
  getIdFromMessage(message, context) {
    const subcommandIdentifier = this.getSubcommandMappingName(context);
    switch (context.scope) {
      case framework.BucketScope.Global:
        return `global.${subcommandIdentifier}`;
      case framework.BucketScope.Channel:
        return `${message.channelId}.${subcommandIdentifier}`;
      case framework.BucketScope.Guild:
        return `${message.guildId}.${subcommandIdentifier}` ?? `${message.channelId}.${subcommandIdentifier}`;
      default:
        return `${message.author.id}.${subcommandIdentifier}`;
    }
  }
  getIdFromInteraction(interaction, context) {
    const subcommandIdentifier = this.getSubcommandMappingName(context);
    switch (context.scope) {
      case framework.BucketScope.Global:
        return `global.${subcommandIdentifier}`;
      case framework.BucketScope.Channel:
        return `${interaction.channelId}.${subcommandIdentifier}`;
      case framework.BucketScope.Guild:
        return `${interaction.guildId}.${subcommandIdentifier}` ?? `${interaction.channelId}.${subcommandIdentifier}`;
      default:
        return `${interaction.user.id}.${subcommandIdentifier}`;
    }
  }
  getSubcommandMappingName(context) {
    return context.subcommandGroupName ? `${context.subcommandGroupName}.${context.subcommandMethodName}` : context.subcommandMethodName;
  }
  getManager(subcommand, context) {
    let manager = this.subcommandBuckets.get(subcommand);
    if (!manager) {
      manager = new ratelimits.RateLimitManager(context.delay, context.limit);
      this.subcommandBuckets.set(subcommand, manager);
    }
    return manager;
  }
};
__name(_PluginPrecondition, "PluginPrecondition");
var PluginPrecondition = _PluginPrecondition;

exports.PluginPrecondition = PluginPrecondition;
