// @sapphire/plugin-subcommands/dist/esm/lib/types/Events.mjs
import '../../chunk-K3DKV2RO.mjs';

// src/lib/types/Events.ts
var SubcommandPluginEvents = {
  ChatInputSubcommandDenied: "chatInputSubcommandDenied",
  ChatInputSubcommandRun: "chatInputSubcommandRun",
  ChatInputSubcommandSuccess: "chatInputSubcommandSuccess",
  ChatInputSubcommandError: "chatInputSubcommandError",
  MessageSubcommandDenied: "messageSubcommandDenied",
  MessageSubcommandRun: "messageSubcommandRun",
  MessageSubcommandSuccess: "messageSubcommandSuccess",
  MessageSubcommandError: "messageSubcommandError",
  SubcommandMappingIsMissingMessageCommandHandler: "subcommandMappingIsMissingMessageCommandHandler",
  SubcommandMappingIsMissingChatInputCommandHandler: "subcommandMappingIsMissingChatInputCommandHandler"
};
var SubcommandPluginIdentifiers = /* @__PURE__ */ ((SubcommandPluginIdentifiers2) => {
  SubcommandPluginIdentifiers2["MessageSubcommandNoMatch"] = "messageSubcommandNoMatch";
  SubcommandPluginIdentifiers2["ChatInputSubcommandNoMatch"] = "chatInputSubcommandNoMatch";
  SubcommandPluginIdentifiers2["SubcommandNotFound"] = "subcommandNotFound";
  return SubcommandPluginIdentifiers2;
})(SubcommandPluginIdentifiers || {});

export { SubcommandPluginEvents, SubcommandPluginIdentifiers };
