// @sapphire/plugin-subcommands/dist/esm/preconditions/PluginSubcommandCooldown.mjs
import { __name, __publicField } from '../chunk-K3DKV2RO.mjs';
import { AllFlowsPrecondition, BucketScope } from '@sapphire/framework';
import { RateLimitManager } from '@sapphire/ratelimits';
import { time, TimestampStyles } from 'discord.js';
import '../lib/Subcommand.mjs';
import { SubcommandIdentifiers } from '../lib/types/Enums.mjs';

var _PluginPrecondition = class _PluginPrecondition extends AllFlowsPrecondition {
  constructor() {
    super(...arguments);
    __publicField(this, "subcommandBuckets", /* @__PURE__ */ new WeakMap());
  }
  messageRun(message, subcommand, context) {
    const cooldownId = this.getIdFromMessage(message, context);
    return this.sharedRun(message.author.id, subcommand, context, cooldownId, "message");
  }
  chatInputRun(interaction, subcommand, context) {
    const cooldownId = this.getIdFromInteraction(interaction, context);
    return this.sharedRun(interaction.user.id, subcommand, context, cooldownId, "chat input");
  }
  contextMenuRun(interaction, subcommand, context) {
    const cooldownId = this.getIdFromInteraction(interaction, context);
    return this.sharedRun(interaction.user.id, subcommand, context, cooldownId, "context menu");
  }
  sharedRun(authorId, subcommand, context, cooldownId, commandType) {
    if (context.external)
      return this.ok();
    if (!context.delay)
      return this.ok();
    if (context.filteredUsers?.includes(authorId))
      return this.ok();
    const rateLimit = this.getManager(subcommand, context).acquire(cooldownId);
    if (rateLimit.limited) {
      const remaining = rateLimit.remainingTime;
      const nextAvailable = time(Math.floor(rateLimit.expires / 1e3), TimestampStyles.RelativeTime);
      return this.error({
        identifier: SubcommandIdentifiers.SubcommandPreconditionCooldown,
        message: `There is a cooldown in effect for this ${commandType} subcommand. It'll be available ${nextAvailable}.`,
        context: { remaining }
      });
    }
    rateLimit.consume();
    return this.ok();
  }
  getIdFromMessage(message, context) {
    const subcommandIdentifier = this.getSubcommandMappingName(context);
    switch (context.scope) {
      case BucketScope.Global:
        return `global.${subcommandIdentifier}`;
      case BucketScope.Channel:
        return `${message.channelId}.${subcommandIdentifier}`;
      case BucketScope.Guild:
        return `${message.guildId}.${subcommandIdentifier}` ?? `${message.channelId}.${subcommandIdentifier}`;
      default:
        return `${message.author.id}.${subcommandIdentifier}`;
    }
  }
  getIdFromInteraction(interaction, context) {
    const subcommandIdentifier = this.getSubcommandMappingName(context);
    switch (context.scope) {
      case BucketScope.Global:
        return `global.${subcommandIdentifier}`;
      case BucketScope.Channel:
        return `${interaction.channelId}.${subcommandIdentifier}`;
      case BucketScope.Guild:
        return `${interaction.guildId}.${subcommandIdentifier}` ?? `${interaction.channelId}.${subcommandIdentifier}`;
      default:
        return `${interaction.user.id}.${subcommandIdentifier}`;
    }
  }
  getSubcommandMappingName(context) {
    return context.subcommandGroupName ? `${context.subcommandGroupName}.${context.subcommandMethodName}` : context.subcommandMethodName;
  }
  getManager(subcommand, context) {
    let manager = this.subcommandBuckets.get(subcommand);
    if (!manager) {
      manager = new RateLimitManager(context.delay, context.limit);
      this.subcommandBuckets.set(subcommand, manager);
    }
    return manager;
  }
};
__name(_PluginPrecondition, "PluginPrecondition");
var PluginPrecondition = _PluginPrecondition;

export { PluginPrecondition };
