// @sapphire/plugin-subcommands/dist/esm/lib/Subcommand.mjs
import { __name, __privateAdd, __publicField, __privateMethod } from '../chunk-K3DKV2RO.mjs';
import { Command, PreconditionContainerArray, PreconditionResolvers, UserError, Result } from '@sapphire/framework';
import { deepClone, cast } from '@sapphire/utilities';
import { SubcommandPreconditionResolvers } from '../index.mjs';
import { SubcommandPluginIdentifiers, SubcommandPluginEvents } from './types/Events.mjs';

var _getMessageParametersAsString, getMessageParametersAsString_fn, _handleMessageRun, handleMessageRun_fn, _handleChatInputInteractionRun, handleChatInputInteractionRun_fn, _findSubcommand, findSubcommand_fn, _supportsCommandType, supportsCommandType_fn;
var _Subcommand = class _Subcommand extends Command {
  constructor(context, options) {
    super(context, options);
    __privateAdd(this, _getMessageParametersAsString);
    __privateAdd(this, _handleMessageRun);
    __privateAdd(this, _handleChatInputInteractionRun);
    __privateAdd(this, _findSubcommand);
    __privateAdd(this, _supportsCommandType);
    /**
     * The preconditions to be run for each specified subcommand.
     * @since 4.1.0
     */
    __publicField(this, "subcommandPreconditions", /* @__PURE__ */ new Map());
    /**
     * The parsed subcommand mappings that were provided with the {@link SubcommandOptions.subcommands} option.
     * This is built at construction time and is used to determine which subcommand to run.
     */
    __publicField(this, "parsedSubcommandMappings");
    /**
     * Whether to use case insensitive subcommands. This is only relevant for message-command styled commands as
     * chat input commands are always lowercase. This can be enabled through `clientOptions.caseInsensitiveCommands`.
     */
    __publicField(this, "caseInsensitiveSubcommands", false);
    this.parsedSubcommandMappings = options.subcommands ?? [];
    const clientOptions = this.container.client.options;
    if (clientOptions.caseInsensitiveCommands) {
      this.caseInsensitiveSubcommands = true;
      for (const cmd of this.parsedSubcommandMappings) {
        cmd.name = cmd.name.toLowerCase();
        if (cmd.type === "group") {
          for (const groupCommand of cmd.entries) {
            groupCommand.name = groupCommand.name.toLowerCase();
          }
        }
      }
    }
    if (options.generateDashLessAliases) {
      for (const mapping of this.parsedSubcommandMappings) {
        if (!Reflect.has(mapping, "messageRun")) {
          continue;
        }
        const dashLessMappings = [];
        if (mapping.type === "group") {
          let hasChangedEntries = null;
          const clonedMapping = deepClone(mapping);
          for (const groupCommand of mapping.entries) {
            if (groupCommand.name.includes("-")) {
              hasChangedEntries ??= true;
              clonedMapping.entries.push({
                ...groupCommand,
                name: groupCommand.name.replaceAll("-", "")
              });
            }
          }
          if (hasChangedEntries) {
            dashLessMappings.push({
              ...mapping,
              entries: clonedMapping.entries
            });
          }
          if (clonedMapping.name.includes("-")) {
            clonedMapping.name = clonedMapping.name.replaceAll("-", "");
            dashLessMappings.push(clonedMapping);
          }
        } else if (mapping.name.includes("-")) {
          dashLessMappings.push({
            ...mapping,
            name: mapping.name.replaceAll("-", "")
          });
        }
        for (const dashLessMapping of dashLessMappings) {
          this.parsedSubcommandMappings.push(dashLessMapping);
        }
      }
    }
    for (const subcommand of this.parsedSubcommandMappings) {
      subcommand.type ??= "method";
      if (subcommand.type === "method") {
        const preconditionContainerArray = new PreconditionContainerArray(subcommand.preconditions);
        PreconditionResolvers.parseConstructorPreConditionsRunIn(
          subcommand.runIn,
          this.resolveConstructorPreConditionsRunType.bind(this),
          preconditionContainerArray
        );
        PreconditionResolvers.parseConstructorPreConditionsNsfw(subcommand.nsfw, preconditionContainerArray);
        PreconditionResolvers.parseConstructorPreConditionsRequiredClientPermissions(
          subcommand.requiredClientPermissions,
          preconditionContainerArray
        );
        PreconditionResolvers.parseConstructorPreConditionsRequiredUserPermissions(
          subcommand.requiredUserPermissions,
          preconditionContainerArray
        );
        SubcommandPreconditionResolvers.parseSubcommandConstructorPreConditionsCooldown({
          subcommand: this,
          cooldownDelay: subcommand.cooldownDelay,
          cooldownFilteredUsers: subcommand.cooldownFilteredUsers,
          cooldownLimit: subcommand.cooldownLimit,
          cooldownScope: subcommand.cooldownScope,
          subcommandMethodName: subcommand.name,
          preconditionContainerArray
        });
        this.subcommandPreconditions.set(subcommand.name, preconditionContainerArray);
      }
      if (subcommand.type === "group") {
        for (const groupedSubcommand of subcommand.entries) {
          const preconditionContainerArray = new PreconditionContainerArray(groupedSubcommand.preconditions);
          PreconditionResolvers.parseConstructorPreConditionsRunIn(
            groupedSubcommand.runIn,
            this.resolveConstructorPreConditionsRunType.bind(this),
            preconditionContainerArray
          );
          PreconditionResolvers.parseConstructorPreConditionsNsfw(groupedSubcommand.nsfw, preconditionContainerArray);
          PreconditionResolvers.parseConstructorPreConditionsRequiredClientPermissions(
            groupedSubcommand.requiredClientPermissions,
            preconditionContainerArray
          );
          PreconditionResolvers.parseConstructorPreConditionsRequiredUserPermissions(
            groupedSubcommand.requiredUserPermissions,
            preconditionContainerArray
          );
          SubcommandPreconditionResolvers.parseSubcommandConstructorPreConditionsCooldown({
            subcommand: this,
            cooldownDelay: groupedSubcommand.cooldownDelay,
            cooldownFilteredUsers: groupedSubcommand.cooldownFilteredUsers,
            cooldownLimit: groupedSubcommand.cooldownLimit,
            cooldownScope: groupedSubcommand.cooldownScope,
            subcommandGroupName: subcommand.name,
            subcommandMethodName: groupedSubcommand.name,
            preconditionContainerArray
          });
          this.subcommandPreconditions.set(`${subcommand.name}.${groupedSubcommand.name}`, preconditionContainerArray);
        }
      }
    }
  }
  onLoad() {
    super.onLoad();
    const externalMapping = Reflect.get(this, "subcommandMappings");
    if (externalMapping) {
      const subcommands = Array.isArray(externalMapping) ? externalMapping : [];
      this.parsedSubcommandMappings = subcommands;
      this.options.subcommands = subcommands;
    }
  }
  /**
   * Whether this command has message-based subcommands or not
   * @returns `true` if this command has message-based subcommands, otherwise `false`
   */
  supportsMessageCommands() {
    return __privateMethod(this, _supportsCommandType, supportsCommandType_fn).call(this, "messageRun");
  }
  /**
   * Whether this command has chat input subcommands or not
   * @returns `true` if this command has chat input subcommands, otherwise `false`
   */
  supportsChatInputCommands() {
    return __privateMethod(this, _supportsCommandType, supportsCommandType_fn).call(this, "chatInputRun");
  }
  /**
   * The method that is ran when a message-based subcommand is ran.
   *
   * **DO NOT** override this in your implementation of a subcommand!
   */
  async messageRun(message, args, context) {
    args.save();
    const subcommandOrGroup = args.nextMaybe();
    const subcommandName = args.nextMaybe();
    let defaultCommand = null;
    let actualSubcommandToRun = null;
    let matchedWithGroupedSubcommand = false;
    for (const mapping of this.parsedSubcommandMappings) {
      mapping.type ??= "method";
      if (mapping.type === "method") {
        if (mapping.default && !defaultCommand) {
          matchedWithGroupedSubcommand = false;
          defaultCommand = mapping;
        }
        if (subcommandOrGroup.isSomeAnd((value) => mapping.name === (this.caseInsensitiveSubcommands ? value.toLowerCase() : value))) {
          actualSubcommandToRun = mapping;
          matchedWithGroupedSubcommand = false;
          break;
        }
      }
      if (mapping.type === "group" && subcommandOrGroup.isSome() && subcommandName.isSome()) {
        const unwrappedSubcommandGroupName = subcommandOrGroup.unwrap();
        const unwrappedSubcommandName = subcommandName.unwrap();
        if (mapping.name === unwrappedSubcommandGroupName) {
          const findResult = __privateMethod(this, _findSubcommand, findSubcommand_fn).call(this, mapping.entries, this.caseInsensitiveSubcommands ? unwrappedSubcommandName.toLowerCase() : unwrappedSubcommandName);
          if (findResult.defaultMatch) {
            defaultCommand = findResult.mapping;
            matchedWithGroupedSubcommand = true;
          } else {
            actualSubcommandToRun = findResult.mapping;
            matchedWithGroupedSubcommand = true;
            break;
          }
        }
      }
    }
    args.restore();
    if (actualSubcommandToRun) {
      args.next();
      let subcommandGroupName = void 0;
      if (matchedWithGroupedSubcommand) {
        subcommandGroupName = subcommandOrGroup.unwrap();
        args.next();
      }
      return __privateMethod(this, _handleMessageRun, handleMessageRun_fn).call(this, message, args, context, actualSubcommandToRun, subcommandGroupName);
    }
    if (defaultCommand) {
      let subcommandGroupName = void 0;
      if (matchedWithGroupedSubcommand) {
        subcommandGroupName = subcommandOrGroup.unwrap();
        args.next();
      }
      return __privateMethod(this, _handleMessageRun, handleMessageRun_fn).call(this, message, args, context, defaultCommand, subcommandGroupName);
    }
    throw new UserError({
      identifier: SubcommandPluginIdentifiers.MessageSubcommandNoMatch,
      message: "No subcommand was matched with the provided arguments.",
      context: {
        ...context,
        possibleSubcommandName: subcommandName.unwrapOr(null),
        possibleSubcommandGroupOrName: subcommandOrGroup.unwrapOr(null)
      }
    });
  }
  /**
   * The method that is ran when a chat input based subcommand is ran.
   *
   * **DO NOT** override this in your implementation of a subcommand!
   */
  async chatInputRun(interaction, context) {
    const subcommandName = interaction.options.getSubcommand(false);
    const subcommandGroupName = interaction.options.getSubcommandGroup(false);
    for (const mapping of this.parsedSubcommandMappings) {
      mapping.type ??= "method";
      if (subcommandGroupName && subcommandName) {
        if (mapping.type !== "group")
          continue;
        if (mapping.name !== subcommandGroupName)
          continue;
        const foundSubcommand = __privateMethod(this, _findSubcommand, findSubcommand_fn).call(this, mapping.entries, subcommandName);
        if (!foundSubcommand.defaultMatch) {
          return __privateMethod(this, _handleChatInputInteractionRun, handleChatInputInteractionRun_fn).call(this, interaction, context, foundSubcommand.mapping, subcommandGroupName);
        }
        continue;
      }
      if (mapping.type === "method" && mapping.name === subcommandName) {
        return __privateMethod(this, _handleChatInputInteractionRun, handleChatInputInteractionRun_fn).call(this, interaction, context, mapping, void 0);
      }
    }
    throw new UserError({
      identifier: SubcommandPluginIdentifiers.ChatInputSubcommandNoMatch,
      message: "No subcommand was matched with the provided command.",
      context
    });
  }
};
_getMessageParametersAsString = new WeakSet();
getMessageParametersAsString_fn = /* @__PURE__ */ __name(async function(args) {
  args.save();
  const parameters = await args.restResult("string");
  args.restore();
  const params = {};
  if (parameters.isOk()) {
    params.parameters = parameters.unwrap();
  }
  return params;
}, "#getMessageParametersAsString");
_handleMessageRun = new WeakSet();
handleMessageRun_fn = /* @__PURE__ */ __name(async function(message, args, context, subcommand, subcommandGroupName) {
  const payload = {
    message,
    command: this,
    context,
    matchedSubcommandMapping: subcommand
  };
  const preconditionsForSubcommand = this.subcommandPreconditions.get(
    subcommandGroupName ? `${subcommandGroupName}.${subcommand.name}` : subcommand.name
  );
  if (preconditionsForSubcommand) {
    const messageParametersAsString = await __privateMethod(this, _getMessageParametersAsString, getMessageParametersAsString_fn).call(this, args);
    const preconditionPayload = { ...messageParametersAsString, ...payload };
    const localSubcommandResult = await preconditionsForSubcommand.messageRun(message, this, preconditionPayload);
    if (localSubcommandResult.isErr()) {
      this.container.client.emit(SubcommandPluginEvents.MessageSubcommandDenied, localSubcommandResult.unwrapErr(), preconditionPayload);
      return;
    }
  }
  const result = await Result.fromAsync(async () => {
    if (subcommand.messageRun) {
      const casted = subcommand;
      this.container.client.emit(SubcommandPluginEvents.MessageSubcommandRun, message, casted, payload);
      let result2;
      if (typeof subcommand.messageRun === "string") {
        const method = Reflect.get(this, subcommand.messageRun);
        if (!method) {
          throw new UserError({
            identifier: SubcommandPluginIdentifiers.SubcommandNotFound,
            message: `The method configured at "messageRun" for the subcommand ${subcommand.name} was not implemented in the class.`,
            context: { ...payload }
          });
        }
        result2 = await Reflect.apply(cast(method), this, [message, args, context]);
      } else {
        result2 = await subcommand.messageRun(message, args, context);
      }
      this.container.client.emit(SubcommandPluginEvents.MessageSubcommandSuccess, message, casted, { ...payload, result: result2 });
    } else {
      this.container.client.emit(SubcommandPluginEvents.SubcommandMappingIsMissingMessageCommandHandler, message, subcommand, payload);
    }
  });
  result.inspectErr((error) => this.container.client.emit(SubcommandPluginEvents.MessageSubcommandError, error, payload));
}, "#handleMessageRun");
_handleChatInputInteractionRun = new WeakSet();
handleChatInputInteractionRun_fn = /* @__PURE__ */ __name(async function(interaction, context, subcommand, subcommandGroupName) {
  const payload = {
    command: this,
    context,
    interaction,
    matchedSubcommandMapping: subcommand
  };
  const preconditionsForSubcommand = this.subcommandPreconditions.get(
    subcommandGroupName ? `${subcommandGroupName}.${subcommand.name}` : subcommand.name
  );
  if (preconditionsForSubcommand) {
    const localSubcommandResult = await preconditionsForSubcommand.chatInputRun(interaction, this, payload);
    if (localSubcommandResult.isErr()) {
      this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandDenied, localSubcommandResult.unwrapErr(), payload);
      return;
    }
  }
  const result = await Result.fromAsync(async () => {
    if (subcommand.chatInputRun) {
      const casted = subcommand;
      this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandRun, interaction, casted, payload);
      let result2;
      if (typeof subcommand.chatInputRun === "string") {
        const method = Reflect.get(this, subcommand.chatInputRun);
        if (!method) {
          throw new UserError({
            identifier: SubcommandPluginIdentifiers.SubcommandNotFound,
            message: `The method configured at "chatInputRun" for the subcommand ${subcommand.name} was not implemented in the class.`,
            context: { ...payload }
          });
        }
        result2 = await Reflect.apply(cast(method), this, [interaction, context]);
      } else {
        result2 = await subcommand.chatInputRun(interaction, context);
      }
      this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandSuccess, interaction, casted, { ...payload, result: result2 });
    } else {
      this.container.client.emit(
        SubcommandPluginEvents.SubcommandMappingIsMissingChatInputCommandHandler,
        interaction,
        subcommand,
        payload
      );
    }
  });
  result.inspectErr((error) => this.container.client.emit(SubcommandPluginEvents.ChatInputSubcommandError, error, payload));
}, "#handleChatInputInteractionRun");
_findSubcommand = new WeakSet();
findSubcommand_fn = /* @__PURE__ */ __name(function(mappings, expectedName) {
  let foundDefault = null;
  for (const mapping of mappings) {
    mapping.type ??= "method";
    if (mapping.default) {
      foundDefault = mapping;
    }
    if (mapping.name === expectedName) {
      return { mapping, defaultMatch: false };
    }
  }
  return { mapping: foundDefault, defaultMatch: true };
}, "#findSubcommand");
_supportsCommandType = new WeakSet();
supportsCommandType_fn = /* @__PURE__ */ __name(function(commandType) {
  return this.parsedSubcommandMappings.some((mapping) => {
    if (mapping.type === "group") {
      return mapping.entries.some((groupCommand) => Reflect.has(groupCommand, commandType));
    }
    return Reflect.has(mapping, commandType);
  });
}, "#supportsCommandType");
__name(_Subcommand, "Subcommand");
var Subcommand = _Subcommand;

export { Subcommand };
